### PR TITLE
refactor: plumb a single context throughout the code

### PR DESCRIPTION
**Pull request by hbagdi**
This patch is an effort to improve the context propagation in decK, down
to every single operation.
In order to do that, a single root context is now being created in the
main function, and all contexts should be derived from it.

### CODE REVIEW

**Review by rainest — DISMISSED** [review]

**Comment by hbagdi** [comment]
Will wait for @mflendrich for a review on this one.

**Comment by mmorel-35** [comment]
The contexts of ping command seems to be missing is it deliberately?

**Comment by hbagdi** [comment]
> The contexts of ping command seems to be missing is it deliberately?

Good catch. Added that.

**Comment by mmorel-35** [comment]
https://github.com/Kong/deck/blob/76d64ae1a05bc1d4611a6dbe58a42c9bf386a78e/cmd/dump.go#L38

**Comment by mmorel-35** [comment]
Shall those lines be removed with this pr?
https://github.com/Kong/deck/blob/64a98cc76ce956ba9d3901e97a211e7744b1f47a/.golangci.yml#L18-L20

**Comment by hbagdi** [comment]
> Shall those lines be removed with this pr?

We can't because solver package has a bunch of nil ctx invocations.
This PR wasn't meant to be an exhaustive squash of all issues with ctx.

**Comment by hbagdi** [comment]
Conflicts have been fixed. PTAL.

File: main.go [hunk 1]
```diff
@@ -11,22 +12,22 @@ import (
...
-		close(done)
+		cancel()
 	}()
+	return ctx
 }
 
 func main() {
-	registerSignalHandler()
-	cmd.Execute()
+	ctx := registerSignalHandler()
```
Comment (mflendrich) [inline main.go:29]: Consider the following:

https://pkg.go.dev/os/signal#NotifyContext

`ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)`

**Review by mflendrich — APPROVED** [review]
I'm happy to see this change :+1:

LGTM! With one optional comment.

File: main.go (hunk 1 above)
Comment (hbagdi) [reply main.go:29]: included this, this is a handy simplification

Comment (hbagdi) [reply main.go:29]: Said too soon, @mflendrich the only problem here is that when the signal arrives, there wouldn't be a message to stdout stating that the signal was actually received. This is quite helpful in debugging. Thoughts?

Comment (mflendrich) [reply main.go:29]: Hacky way (not recommended):

```go
go func() {
  select {
    case <-ctx.Done():
      fmt.Printf("context cancelled (%w), terminating...", ctx.Err())
  }
}
```

Clean way:
- Replace `os.Exit(1)` throughout the codebase with errors plumbed all the way through to `RunE` and to the return value of `cmd.Execute()`
- After `cmd.Execute()` returns, check if `ctx.Err()` is not nil. If it isn't, then the context got cancelled by the signal.

Both are somewhat tedious and may defeat the purpose of this simplification.

Comment (hbagdi) [reply main.go:29]: Thought and explored a few different ways to do it but I haven't found a good way to deal with it in this PR.
This also feels minor enough that we can leave it be for now. If it is a bigger issue, please LMK and I'll address with a quick follow up PR.
I'm going to go ahead and merge this in.

**Comment by mmorel-35** [comment]
Part 1of #325

**Pull request merged**